### DISCUSSION TITLE
[ Common ] Increasing display resolution without increasing internal resolution

### DISCUSSION THREAD

**Discussion by JordanAnthonyKing** [julianxhokaxhiu/FFNx#297] [Q&A] [answered]
## Summary

Hi, I can't see if this feature is already supported by some means, I can't find anything in the 7th Heaven driver settings.

Is it possible to render the graphics at a lower resolution than the output resolution for a crisp, pixelated look without changing the display resoltion? Preferably with the ability to render internally at say 8x, downsample to 1x, then display at 8x without filtering.

### Motivation

That retro feeling: using crt-royale with reshade requires the display resolution to still be high, even if the game resolution isn't.

**Comment by julianxhokaxhiu** [comment, marked as answer]
Hi, thanks for having me approached.

Anything about 7th-Heaven needs to be discussed in their relative support channels, which this one is not the place.

Nevertheless, you can absolutely achieve that manually editing the `FFNx.toml` file at this entry: https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.toml#L43

Best regards.

**Reply by JordanAnthonyKing** [reply]
Hi, thanks for getting back to me, I did notice this option, but I didn't see how this would give the desired effect. Sure, I could run at 640x480, with the internal resolution at 1280x960, but then how would I integer scale the display back up to 1280x960?

**Reply by julianxhokaxhiu** [reply]
Probably I'm missing something in this picture but I think you're just looking at what is being already done, just using another mental flow but FFNx is already doing that:

1) You set your output resolution, so FFNx knows which is your target ( say you use 640x480 so it's the 1x you mentally think )
2) You set your `internal_resolution_scale = 8` ( so it's the 8x in your step )
3) FFNx will resize back to your resolution ( what you think as 8x, but you can't output a bitmap bigget than your resolution is, otherwise you'll see just a portion of it, not the entire picture )

If you don't want any filter on step 3, set `enable_antialiasing = 0` and that will get your desired "without filtering" effect.

**Reply by JordanAnthonyKing** [reply]
I don't appear to be getting the desired effect.

Here's the relevant section of my FFNx.toml:
```
#[RESOLUTION]
# Resolution of the game.
# Default (value = 0):
# - Window mode will use 640x480
# - Fullscreen mode will use your desktop resolution
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
window_size_x = 320
window_size_y = 240

#[INTERNAL RESOLUTION SCALE]
# The scale is in multiples of 640x480
# The scale factor is used to render internally at a larger size then the display, this will then be downscaled to the games current resolution
# This is required to avoid visual glitches that may happen when the game is not rendered in a 4:3 aspect ratio.
# This is optional for native 4:3 resolutions, but it will enhance visual quality ( like Nvidia DSR ).
# Default = 2
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
internal_resolution_scale = 8

#[PRESERVE ASPECT]
# Preserve original game aspect ratio of (4:3) by adding black bars on the left and right side (if needed)
# When off the game will be stretched to fit the window's aspect ratio; Be aware the game may look wrong though.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
preserve_aspect = true

#[ENABLE VSYNC]
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
enable_vsync = false

#[REFRESH RATE]
# Screen refresh rate.
# Default is 0 = use current screen refresh rate
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
refresh_rate = 0

#[ANTIALIASING]
# Enable antialiasing filtering, this is done in the last pass when downsampling from the current supersampled rendering ( based on internal_resolution_scale logic )
# Available choices are:
# - 0: Disabled ( Default )
# - 2: 2x
# - 4: 4x
# - 8: 8x
# - 16: 16x
# If you set any value that is different than those, it will pick the higher option ( eg. if you pick 6 it will be 8, if you pick 1 it will be 2, and so on ).
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
enable_antialiasing = 0
```

This results in, by default a very small window (as expected), but resizing to fill my monitor still results in nice, crisp visuals:
![image](https://user-images.githubusercontent.com/58787893/139415322-828fa3a4-7c83-47d4-b900-bf47b454d853.png)

I have mods enabled, but would still expect the entire image at this res to be pixelated, which it isn't. The same is true without mods:
![image](https://user-images.githubusercontent.com/58787893/139416028-5d453598-6187-4123-99e0-385e7758104f.png)


Am I misunderstanding a setting?

Doing this also doesn't work with fullscreen, giving the error 'unable to enable the desired resolution', as the game is setting my display resolution, instead of just upscaling the image to it as desired.

**Reply by julianxhokaxhiu** [reply]
There's no "stretched upscaling" logic implemented at all, sorry. If you want to do that feel free to use Reshade ( https://reshade.me/ ) on top of FFNx.

When you maximize the window, it will update the resolution set ( 320x240 in your case ) to the window resolution size. That's why you still get it crisp. The same it will happen when going Fullscreen.

**Reply by JordanAnthonyKing** [reply]
Okay, thanks for helping out.

**Reply by julianxhokaxhiu** [reply]
> Okay, thanks for helping out.

You're welcome! Sorry I can't help but I don't see this as a valid use case. The idea behind the mod is to make the game look as great as possible on any new hardware. Doing the contrary is really beyond the scope.

Cheers!

**Reply by JordanAnthonyKing** [reply]
Hi again, I’ve done some fiddling and couldn’t find a solution I liked with downscaling the output image.

If I were to fork the project (just for myself) and try implementing this feature, could you give me a pointer on a good place to start?

**Comment by chrcoluk** [comment]
Looking at the documentation of the toml, the scale is in multiple of 640x480, assuming 1 is the lowest setting, then the lowest rendering resolution is 640x480.

So from my understanding.

1 - Rendering resolution is always a multiple of 640x480.
2 - 640x480 is the lowest rendering resolution possible.
3 - The window size resolution is just the output resolution of which it will scale to, and thats why it didnt work as you expected.

If you want to render at 320x240, then perhaps you need to request for the internal resolution to work with a scale of 0.5, or that it uses a multiple of 320x240 instead of 640x480.

So if e.g. you want a 2560x1440 output resolution, but most retro look as possible, then you would use these values.

Window resolution 2560x1440
Full screen on
Internal resolution scale 1
Preserve aspect on
